src/rust-crypto/backup.ts

```diff
@@ -591,6 +603,173 @@ export class RustBackupManager extends TypedEventEmitter<RustBackupCryptoEvents,
+    ): Promise<KeyBackupRestoreResult> {
+        const keyBackup = await this.downloadKeyBackup(backupInfoVersion);
+        opts?.progressCallback?.({
+            stage: "load_keys",
+        });
+
+        return this.importKeyBackup(keyBackup, backupInfoVersion, backupDecryptor, opts);
+    }
+
+    /**
+     * Call `/room_keys/keys` to download the key backup (room keys) for the given backup version.
+     * https://spec.matrix.org/latest/client-server-api/#get_matrixclientv3room_keyskeys
+     *
+     * @param backupInfoVersion
+     * @returns The key backup response.
+     */
+    private downloadKeyBackup(backupInfoVersion: string): Promise<KeyBackup> {
+        return this.http.authedRequest<KeyBackup>(
+            Method.Get,
+            "/room_keys/keys",
+            { version: backupInfoVersion },
+            undefined,
+            {
+                prefix: ClientPrefix.V3,
+            },
+        );
+    }
+
+    /**
+     * Import the room keys from a `/room_keys/keys` call.
+     * Call the opts.progressCallback with the progress of the import.
+     *
+     * @param keyBackup - The response from the server containing the keys to import.
+     * @param backupInfoVersion - The version of the backup info.
+     * @param backupDecryptor - The backup decryptor to use to decrypt the keys.
+     * @param opts - Options for the import.
+     *
+     * @returns The total number of keys and the total imported.
+     *
+     * @private
+     */
+    private async importKeyBackup(
+        keyBackup: KeyBackup,
+        backupInfoVersion: string,
+        backupDecryptor: BackupDecryptor,
+        opts?: KeyBackupRestoreOpts,
+    ): Promise<KeyBackupRestoreResult> {
+        // We have a full backup here, it can get quite big, so we need to decrypt and import it in chunks.
+
+        // Get the total count as a first pass
+        const totalKeyCount = this.getTotalKeyCount(keyBackup);
+        let totalImported = 0;
+        let totalFailures = 0;
+
+        try {
+            // Now decrypt and import the keys in chunks
+            await this.handleDecryptionOfAFullBackup(keyBackup, backupDecryptor, 200, async (chunk) => {
+                // We have a chunk of decrypted keys: import them
+                try {
+                    await this.importBackedUpRoomKeys(chunk, backupInfoVersion);
+                    totalImported += chunk.length;
+                } catch (e) {
+                    totalFailures += chunk.length;
+                    // We failed to import some keys, but we should still try to import the rest?
+                    // Log the error and continue
+                    logger.error("Error importing keys from backup", e);
+                }
+
+                opts?.progressCallback?.({
+                    total: totalKeyCount,
+                    successes: totalImported,
+                    stage: "load_keys",
+                    failures: totalFailures,
+                });
+            });
```

Comment:
Given that `handleDecryptionOfAFullBackup` is only called in one place, this callback structure feels a bit unncessary. Couldn't you just inline `handleDecryptionOfAFullBackup` ?